### ISSUE TITLE
Migrating from old OIDC to new OIDC provider configuration breaks authentication for existing accounts

### ISSUE THREAD

**Issue by luilegeant** [matrix-org/synapse#9514]
### Description

When upgrading from Synapse 1.26.0 to 1.27.0 or 1.28.0 (I can't figure out when the issue came up) the configuration for the OIDC provider (keycloak in my case) was changed as instructed on the upgrade file in release notes.
Since then, when i try to log in from a new device, a new account is created, instead of heaving `@user:domain.tld` I get `@user1:domain.tld` (notice the 1).

### Steps to reproduce

- Setup synapse 1.26.0 with the "old" way of configuring the oidc provider.
- Log in with one of your new account (you user gets created `@user:domain.tld`)
- Update to synapse 1.27.0 or 1.28.0 (I forgot which requires the configuration to be changed) & adapt your oidc configuration
- Log in with the same credential on a new device (old device still have an old valid state, so they continue working).
- When confirmation prompt will be `@user1:domain.tld` (notice the 1).

Expected behaviour:

- I can continue logging in with the same credential without triggering the creation of a new account.

### Version information

Homeserver with the issue identified on:
- `matrix.luilegeant.com`
- `matrix.torquetum.eu`

 What version of Synapse is running?

- **Version**: `{"server_version":"1.28.0","python_version":"3.8.8"}`

- **Install method**:
docker (image see: https://hub.docker.com/r/matrixdotorg/synapse)

- **Platform**:
docker container running in swarm mode on Ubuntu Server 18.04 LTS behind a reverse proxy.

### Extra notes:

- Upon reading release notes and digging a little into the repository, I came upon this [note at Line 73](https://github.com/matrix-org/synapse/blob/0d81a6fa3e1dc832f56ed09805229b9089758ba5/synapse/config/oidc_config.py#L73).
Could it be linked, because the id changed from the old config style to the new one ?
- Second note: would the configuration attribute [described on line 129](https://github.com/matrix-org/synapse/blob/0d81a6fa3e1dc832f56ed09805229b9089758ba5/synapse/config/oidc_config.py#L129) resolve my issue (temporary fix)

**Comment by richvdh** [comment]
can you clarify how you've changed your Synapse OIDC configuration? What was it before, and what is it now?

The problem is almost certainly related to `idp_id`, but the old style of config should work fine with no changes.

**Comment by luilegeant** [comment]
#### Complementary note:
I've made a mistake, its not 1.26.0 configuration style but **1.25.0** (although I've changed only when switching from 1.26.0 upward and needed to login when I was already on 1.28.0).

#### Case 1: reverting to old style configuration (element-web not happy)
My old configuration was (recomposition, I don't have a backup that old):
```yaml
oidc_config:
    enable: true
    issuer: "myissueruri"
    client_id: "synapse"
    client_secret: "mysecret"
    scopes: ["openid", "profile"]
    user_mapping_provider:
      config:
        localpart_template: "{{ user.preferred_username }}"
        display_name_template: "{{ user.given_name }}"
```
If I try using the above configuration element web (1.21.0) on my synapse (1.28.0) and element-web says _"This homeserver doesn't offer any login flows which are supported by this client."_

#### Case 2: using new style configuration (creates new user)
The new configuration:
```yaml
oidc_providers:
  - idp_id: keycloak
    idp_name: "A fancy name"
    issuer: "myissueruri"
    client_id: "synapse"
    client_secret: "mysecret"
    scopes: ["openid", "profile"]
    user_mapping_provider:
      config:
        localpart_template: "{{ user.preferred_username }}"
        display_name_template: "{{ user.given_name }}"
```
This is the case described above in the head of the ticket.

**Comment by richvdh** [comment]
>     enable: true

this should be `enabled`. I think this is the reason the old config doesn't work.

If you do want to switch to the new style, you'll need to set the `idp_id` to `oidc` to keep the old user mappings, as documented at https://github.com/matrix-org/synapse/blob/master/docs/sample_config.yaml#L1849-L1851.

**Comment by luilegeant** [comment]
Hello again,
Thank you very much, this is the part that I missed, using the new configuration format and using `oidc` as `ipd_id` resolved my problems.

Summary of my story:
- When upgrading from 1.27.0 to 1.26.0 
- I've read the [upgrade.rst](https://github.com/matrix-org/synapse/blob/v1.27.0/UPGRADE.rst) (warning version 1.27.0 tagged in this link)
- Then ended reviewing the [docs/openid.md](https://github.com/matrix-org/synapse/blob/v1.27.0/docs/openid.md)  (warning version 1.27.0 tagged in this link)
- I skipped reviewing the sample config linked in the page, i only looked at the quoted configration => this is where i failed. :facepalm: 
Side note: The problem seemed to have ocured on 1.28.0 because i've only need to log in again when i was running this version.

**Nice to have**: in the `docs/openid.md` or `upgrade.rst` files: the quote you've mentioned above that resolved everything: https://github.com/matrix-org/synapse/blob/master/docs/sample_config.yaml#L1849-L1851

Thank you very much for your help !

**Comment by luilegeant** [comment]
Closing since the error is on my side (configuration) and not code wise.

**Comment by richvdh** [comment]
glad you got it solved. One more thing: you said:

> the configuration for the OIDC provider (keycloak in my case) was changed as instructed on the upgrade file in release notes.

can you clarify which bit of the release notes/`UPGRADE.rst` made you think that changing the synapse configuration was required? It would be helpful to know so that we can avoid making similar mistakes in future?

**Comment by luilegeant** [comment]
It's not written per se, I'll describe step by step how i lead myself into the rabbit hole :rabbit2: 
- The `UPGRADE.rst` file mentions _"If your server is configured for single sign-on via an OpenID Connect or OAuth2 identity provider, you will need to add (...) to the list of permitted "redirect URIs" at the identity provider."_ then continues by linking to `docs/openid.md`.
- I went on and read the document.
- Skipped the sample config file => first mistake (which mention the oidc trick you found and resolved my issue)
- Reading further on, I realised the configuration file changed on the oidc section configuration
- I assumed i should align (to be able to continue working). => second mistake

For me what is missing:
- mention that the old config format still works (until when ?)
- the quote in the [sample config file](https://github.com/matrix-org/synapse/blob/master/docs/sample_config.yaml#L1849-L1851) could be copied into the `docs/openid.md`.
